File: dcp-solutions/2024-10-15/solution.rb
```ruby
# Definition for a binary tree node.
class TreeNode
  attr_accessor :val, :left, :right
  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

def second_largest_node(root)
  max = -Float::INFINITY
  max2 = -Float::INFINITY
  traverse(root) do |node|
    return node.val if node.val > max
    (max2 = node.val) and break if node.val > max2 && node.val != max
    max = node.val
  end
  max2 == -Float::INFINITY ? nil : max2
end

def traverse(node, &block)
  return unless block_given?
  block.call(node) if node
  traverse(node.left, &block) if node
  traverse(node.right, &block)
end
```

File: dcp-solutions/2024-10-15/solution.go
```markdown
package main

import (
    "errors"
    "fmt"
)

// Definition for a binary tree node.
type TreeNode struct {
    Val   int
    Left  *TreeNode
    Right *TreeNode
}

func secondHighest(root *TreeNode) (int, error) {
    if root == nil {
        return 0, errors.New("root is null")
    }

    leftMax := make(map[int]int)
    rightMax := make(map[int]int)

    var f func(*TreeNode)
    f = func(node *TreeNode) {
        if node == nil {
            return
        }
        rightMax[node.Val]++
        f(node.Left)
        f(node.Right)
    }

    f(root)

    maxFromLeft, countFromLeft := 0, 0
    for v, c := range leftMax {
        if countFromLeft < 2 {
            maxFromLeft = v
            countFromLeft += c
        } else if c > 0 {
            return v, errors.New("more than one maximum value")
        }
    }

    maxFromRight, countFromRight := 0, 0
    for v, c := range rightMax {
        if countFromRight < 2 {
            maxFromRight = v
            countFromRight += c
        } else if c > 0 {
            return -1, errors.New("more than one maximum value")
        }
    }

    // select the higher of two numbers.
    result := max(maxFromLeft, maxFromRight)
    if result == 0 {
        return 0, errors.New("no node with more than a single child exists")
    } else {
        return result, nil
    }
}
```